### ISSUE TITLE
warning: calling a __host__ function from a __host__ __device__ function and runtime errors

### ISSUE THREAD

**Issue by anuppari** [thrust/thrust#934]
### Issue
Simple use of `thrust::remove_if` causes compile time warnings and runtime errors. At compile time, I get the following warning:
```bash
warning: calling a __host__ function("thrust::detail::aligned_reinterpret_cast<int *, void *> ") from a __host__ __device__ function("thrust::cuda_cub::__copy_if::copy_if< ::thrust::cuda_cub::par_t,  ::thrust::device_ptr<int> ,  ::thrust::cuda_cub::__copy_if::no_stencil_tag_ *,  ::thrust::device_ptr<int> ,  ::thrust::detail::unary_negate< ::is_even> > ") is not allowed
```
### Minimum working example
My MWE is below, which is basically the same as the [example](https://thrust.github.io/doc/group__stream__compaction.html#ga307d7f64566909172a3f9e16b7e2ad53) but with device memory.

```C++
#include <iostream>
#include <thrust/device_vector.h>
#include <thrust/host_vector.h>
#include <thrust/execution_policy.h>
#include <thrust/remove.h>

// Basic predicate
struct is_even
{
  __host__ __device__
  bool operator()(const int x)
  {
    return (x % 2) == 0;
  }
};

int main()
{
  // Allocate gpu memory and initialize
  const int N = 6;
  int *arrayPtr, *sizePtr;
  cudaMalloc((void **)&arrayPtr, sizeof(int)*N);
  cudaMalloc((void **)&sizePtr, sizeof(int));
  cudaMemset(arrayPtr, 0, sizeof(int)*N);
  cudaMemcpy(sizePtr, &N, sizeof(int), cudaMemcpyHostToDevice);
  
  // Wrap with thrust pointers
  thrust::device_ptr<int> thrustSizePtr(sizePtr);
  thrust::device_ptr<int> thrustArrayPtr(arrayPtr);
  std::cout << thrustSizePtr << " " << *thrustSizePtr << std::endl;

  // Thrust algorithm
  *thrustSizePtr = thrust::distance(thrustArrayPtr, thrust::remove_if(thrust::device, thrustArrayPtr, thrust::device_ptr<int>(arrayPtr+N), is_even()));

  // Output
  std::cout << thrustSizePtr << " " << *thrustSizePtr << std::endl;

  // Free memory
  cudaFree(arrayPtr);
  cudaFree(sizePtr);

  return 0;
}
```
At runtime, this produces the following output:
```
0x7fffc7600200 6
0x7fffc7600200 -128
```
whereas I would expect the new array size to be non-negative.

I'm building using CMake, with the following `CMakeLists.txt`

```cmake
cmake_minimum_required(VERSION 3.9)
project(test_thrust LANGUAGES CXX CUDA)

find_package(CUDA REQUIRED)

add_executable(test "test.cu")

target_compile_features(test PUBLIC cxx_std_11)
set_target_properties(test PROPERTIES CUDA_SEPARABLE_COMPILATION ON)
set_target_properties(test PROPERTIES POSITION_INDEPENDENT_CODE ON)
set(CMAKE_CUDA_FLAGS "-arch=sm_61 -Xcompiler -rdynamic -G")

target_link_libraries(test cuda)
```

### System:
OS: Ubuntu 16.04 64-bit
Kernel version: 4.15.0-36
Cuda version: Cuda 10
Thrust version: 1.9.3 (version included with cuda)
Driver version: 410.48
GPU: GTX 1080

**Comment by griwes** [comment]
```
  *thrustSizePtr = thrust::distance(thrustSizePtr, thrust::remove_if(thrust::device, thrustArrayPtr, thrust::device_ptr<int>(arrayPtr+N), is_even()));
```

Did you mean this instead?

```
  *thrustSizePtr = thrust::distance(thrustArrayPtr, thrust::remove_if(thrust::device, thrustArrayPtr, thrust::device_ptr<int>(arrayPtr+N), is_even()));
```

**Comment by anuppari** [comment]
The warnings only seem to show up when compiled with `-arch=sm_35` or later. For `-arch=sm_30`, there are no warnings.

**Comment by griwes** [comment]
This has been fixed on master with a6d41f66efde6bf8d7220aa6013b5ffa8cd95122.

**Comment by ENM185** [comment]
@griwes is there a way to suppress the warning (specifically on aligned_reinterpret_cast) in CUDA 10.0? (and is it safe to do so?)
